lib/Serialization/ModuleFileSharedCore.cpp

```diff
@@ -472,13 +470,31 @@ std::string ModuleFileSharedCore::Dependency::getPrettyPrintedPath() const {
   return output;
 }
 
-void ModuleFileSharedCore::fatal(llvm::Error error) {
-  logAllUnhandledErrors(std::move(error), llvm::errs(),
-                        "\n*** DESERIALIZATION FAILURE (please include this "
-                        "section in any bug report) ***\n");
+void ModuleFileSharedCore::fatal(llvm::Error error) const {
+  llvm::SmallString<0> errorStr;
+  llvm::raw_svector_ostream out(errorStr);
+
+  out << "*** DESERIALIZATION FAILURE ***\n";
+  outputDiagnosticInfo(out);
+  out << "\n";
+  if (error) {
+    handleAllErrors(std::move(error), [&](const llvm::ErrorInfoBase &ei) {
+      ei.log(out);
+      out << "\n";
+    });
+  }
+
+  llvm::PrettyStackTraceString trace(errorStr.c_str());
   abort();
 }
 
+void ModuleFileSharedCore::outputDiagnosticInfo(llvm::raw_ostream &os) const {
+  os << "module '" << Name << "' with full misc version '" << MiscVersion
+      << "'";
+  if (Bits.IsAllowModuleWithCompilerErrorsEnabled)
+    os << " (built while allowing compiler errors)";
```

Comment:
Could we use the flag instead here `built with -experimental-allow-module-with-compiler-errors`? It would be more precise and we could add more relevant flags to it in the future.

Reply:
Sure, sounds good 👍